Spigot-Server-Patches/0581-Improve-ServerGUI.patch

```diff
@@ -0,0 +1,121 @@
+ import java.awt.Dimension;
+ import java.awt.Graphics;
+@@ -16,7 +18,7 @@ public class GuiStatsComponent extends JComponent {
+     });
+     private final int[] b = new int[256];
+     private int c;
+-    private final String[] d = new String[11];
++    private final String[] d = new String[12];
+     private final MinecraftServer e;
+     private final Timer f;
+ 
+@@ -35,8 +37,13 @@ public class GuiStatsComponent extends JComponent {
+     private void b() {
+         long i = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+ 
++        // Paper start - Add tps entry
++        double tps = Bukkit.getTPS()[0];
+         this.d[0] = "Memory use: " + i / 1024L / 1024L + " mb (" + Runtime.getRuntime().freeMemory() * 100L / Runtime.getRuntime().maxMemory() + "% free)";
+         this.d[1] = "Avg tick: " + GuiStatsComponent.a.format(this.a(this.e.h) * 1.0E-6D) + " ms";
++        this.d[2] = "TPS:" + ( ( tps > 21.0 ) ? "*" : "" ) + Math.min( Math.round( tps * 100.0 ) / 100.0, 20.0 );
++        // Paper end
++
+         this.b[this.c++ & 255] = (int) (i * 100L / Runtime.getRuntime().maxMemory());
+         this.repaint();
+     }
+diff --git a/src/main/java/net/minecraft/server/ServerGUI.java b/src/main/java/net/minecraft/server/ServerGUI.java
+index e51434f2323bcb9385d11e99d0438d66434e3e63..d2184e79dd3074da41565be32d5b6f5415307688 100644
+--- a/src/main/java/net/minecraft/server/ServerGUI.java
++++ b/src/main/java/net/minecraft/server/ServerGUI.java
+@@ -10,7 +10,9 @@ import java.awt.event.FocusAdapter;
+ import java.awt.event.FocusEvent;
+ import java.awt.event.WindowAdapter;
+ import java.awt.event.WindowEvent;
++import java.io.IOException;
+ import java.util.Collection;
++import java.util.Objects;
+ import java.util.concurrent.atomic.AtomicBoolean;
+ import javax.swing.JComponent;
+ import javax.swing.JFrame;
+@@ -22,6 +24,9 @@ import javax.swing.JTextArea;
+ import javax.swing.JTextField;
+ import javax.swing.SwingUtilities;
+ import javax.swing.UIManager;
++import javax.swing.WindowConstants;
++import javax.swing.ScrollPaneConstants;
++import javax.imageio.ImageIO;
+ import javax.swing.border.EtchedBorder;
+ import javax.swing.border.TitledBorder;
+ import javax.swing.text.AttributeSet;
+@@ -32,7 +37,7 @@ import org.apache.logging.log4j.Logger;
+ 
+ public class ServerGUI extends JComponent {
+ 
+-    private static final Font a = new Font("Monospaced", 0, 12);
++    private static final Font a = new Font("Monospaced", Font.PLAIN, 12); // Paper
+     private static final Logger LOGGER = LogManager.getLogger();
+     private final DedicatedServer c;
+     private Thread d;
+@@ -49,15 +54,25 @@ public class ServerGUI extends JComponent {
+         final JFrame jframe = new JFrame("Minecraft server");
+         final ServerGUI servergui = new ServerGUI(dedicatedserver);
+ 
+-        jframe.setDefaultCloseOperation(2);
++        jframe.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE); // Paper
```

Comment:
No need changing the constant, I can't see why we'd want a conflict on it if it ever changes.